### DISCUSSION TITLE
Fixing 'Import cycle from Django settings module prevents type inference'

### DISCUSSION THREAD

**Discussion by intgr** [typeddjango/django-stubs#2097] [Common issues and solutions]
In recent versions, the mypy plugin of django-stubs has started issuing a new warning in cases where the plugin is unable to resolve the type of a setting, for example:

```python
from django.conf import settings

print(settings.SOME_SETTING)
```

Causes error:

> error: Import cycle from Django settings module prevents type inference for 'SOME_SETTING'  [misc]

The recommended solution is to add an explicit type hint into the main settings file that is used by the django-stubs:

```diff
-SOME_SETTING = func_returning_int()
+SOME_SETTING: int = func_returning_int()
```

(thanks to @flaeppe in https://github.com/typeddjango/django-stubs/pull/1163#issuecomment-1397544906)

**Comment by flaeppe** [comment]
I'm honestly unsure why this error is emitted by the plugin in the first place.

Deserves to be investigated if it's a remnant of an old mypy crash fix, or if it leads to any other issue(s).

I'm suspecting that we might be able to remove the code yielding the error

**Reply by intgr** [reply]
I opened a PR with a test case that demonstrates this in practice. #2098

**Comment by shmulvad** [comment]
I'm seeing this error as well after upgrading to 5.0.0, in some cases for extremely obvious variables where it should be trivial to infer the type. For example:

```python
# settings.py
DEFAULT_FROM_EMAIL = f'Company Name <mail@companyname.com>'
```

```bash
$ mypy .
# error: Import cycle from Django settings module prevents type inference for 'DEFAULT_FROM_EMAIL'  [misc]
```

If I explicitly add the type hint, it works

```python
# settings.py
DEFAULT_FROM_EMAIL: str = f'Company Name <mail@companyname.com>'
```

```bash
$ mypy .
# Success: no issues found in [x] source files
```

Of course, I can make the error go away by adding the type hint, but I'd prefer to not have to explicitly add types for all variables in my `settings.py` that are used elsewhere and can easily be inferred.

**Reply by intgr** [reply]
Indeed, I had 2 new cases of this error show up as well with the upgrade to 5.0.0, even though this "import cycle" logic was already present in the previous version (and I think a few versions back already).

I guess there was a regression related to settings resolution in 5.0.0.

**Reply by flaeppe** [reply]
Perhaps you could try to bisect the 2 versions? See if we could pin point what change affected the settings resolution?

**Reply by intgr** [reply]
Indeed, "[e517a1fb2bcdad20cc50930c43b285cb0650f292](https://github.com/typeddjango/django-stubs/commit/e517a1fb2bcdad20cc50930c43b285cb0650f292) is the first bad commit", it comes from PR #2024

And reverting that commit removes the mypy errors I was experiencing.

**Reply by flaeppe** [reply]
Interesting, perhaps we should change the priority only for `django.conf.settings`?

**Reply by flaeppe** [reply]
I even think it's reasonable to use `PRI_HIGH` for `django.conf.settings` since that's essentially a proxy for importing the real settings module.

**Reply by flaeppe** [reply]
I just tried it out and realised that increasing to `PRI_HIGH` the test case added in #2098 breaks. The error is not emitted any longer.

**Reply by intgr** [reply]
I opened a PR as well. Anyway let's continue the discussion there.
